flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java

```diff
@@ -0,0 +1,383 @@
+
+    protected final transient DataType valueType;
+    protected final transient DataType frequencyType;
+
+    public PercentileAggFunction(LogicalType inputType, LogicalType frequencyType) {
+        this.valueType = toInternalDataType(inputType);
+        this.frequencyType = frequencyType == null ? null : toInternalDataType(frequencyType);
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Accumulator
+    // --------------------------------------------------------------------------------------------
+
+    /** Accumulator for PERCENTILE. */
+    public static class PercentileAccumulator {
+
+        public double[] percentages;
+        public long totalCount;
+        public MapView<Double, Long> valueCount;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            PercentileAccumulator that = (PercentileAccumulator) o;
+            return Arrays.equals(percentages, that.percentages)
+                    && totalCount == that.totalCount
+                    && Objects.equals(valueCount, that.valueCount);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(Arrays.hashCode(percentages), totalCount, valueCount.hashCode());
+        }
+
+        public Double[] getValue() {
+            // calculate the position for each percentage and sort it, so that all percentiles can
+            // be obtained with a single traversal
+            List<Pair<Double, Integer>> sortedPercentages = new ArrayList<>();
+            for (int index = 0; index < percentages.length; index++) {
+                sortedPercentages.add(new Pair<>(percentages[index] * (totalCount - 1) + 1, index));
+            }
+            sortedPercentages.sort(Comparator.comparing(Pair::getKey));
+
+            // get a sorted list from hashmap for single traversal
+            List<Map.Entry<Double, Long>> sortedList = new ArrayList<>();
+            try {
+                for (Map.Entry<Double, Long> entry : valueCount.entries()) {
+                    sortedList.add(entry);
+                }
+            } catch (Exception e) {
+                throw new FlinkRuntimeException(e);
+            }
+            sortedList.sort(Map.Entry.comparingByKey());
+
+            Double[] percentiles = new Double[percentages.length];
+
+            long preCnt = sortedList.get(0).getValue();
+            for (int i = 0, j = 0; i < sortedPercentages.size(); i++) {
+                Pair<Double, Integer> entry = sortedPercentages.get(i);
+                double position = entry.getKey();
+                long lower = (long) Math.floor(position);
+                long higher = (long) Math.ceil(position);
+
+                // lower <= (totalCount - 1) * percentage + 1 <= totalCount
```

Comment:
This holds true when state ttl is not set.
However, once ttl is set, the `totalCount` and `valueCount` states in acc are not atomic, and there may be corner cases where `totalCount` still exists but the `valueCount` state has expired, in which case the `totalCount` value may be much larger than the actual state of `valueCount`

Reply:
`totalCount` has been removed from accumulator in the latest commit, and will be calculated in `getValue()` once needed.